pyuvsim/uvsim.py

```diff
@@ -360,7 +360,251 @@ def _check_ntasks_valid(Ntasks_tot):
+    Nsky_parts,
+    Nbls,
+    Ntimes,
+    Nfreqs,
+    Nsrcs,
+    quiet,
+):
+    mpi.start_mpi()
+    rank = mpi.get_rank()
+    comm = mpi.get_comm()
+
+    if rank == 0:
+        vis_data = mpi.MPI.Win.Create(uv_container._data_array.value, comm=mpi.world_comm)
+        pbar = simutils.progsteps(maxval=Ntasks_tot)
+    else:
+        vis_data = mpi.MPI.Win.Create(None, comm=mpi.world_comm)
+
+    engine = UVEngine()
+    count = mpi.Counter()
+    size_complex = np.ones(1, dtype=complex).nbytes
+    data_array_shape = (Nbls * Ntimes, 1, Nfreqs, 4)
+    uvdata_indices = []
+
+    for task in local_task_iter:
+        engine.set_task(task)
+        vis = engine.make_visibility()
+
+        blti, spw, freq_ind = task.uvdata_index
+
+        uvdata_indices.append(task.uvdata_index)
+
+        flat_ind = np.ravel_multi_index(
+            (blti, spw, freq_ind, 0), data_array_shape
+        )
+        offset = flat_ind * size_complex
+
+        vis_data.Lock(0)
+        vis_data.Accumulate(vis, 0, target=offset, op=mpi.MPI.SUM)
+        vis_data.Unlock(0)
+
+        cval = count.next()
+        if rank == 0 and not quiet:
+            pbar.update(cval)
+
+    request = comm.Ibarrier()
+
+    while not request.Test():
+        if rank == 0 and not quiet:
+            cval = count.current_value()
+            pbar.update(cval)
+
+    count.free()
+    if rank == 0 and not quiet:
+        pbar.finish()
+
+    if rank == 0 and not quiet:
+        print("Calculations Complete.", flush=True)
+
+    vis_data.Free()
+    if rank == 0:
+        return uv_container, uvdata_indices
+    return None, uvdata_indices
+
+
+def _run_uvsim_tqdm(
+    uv_container,
+    local_task_iter,
+    Ntasks_tot,
+    Nsky_parts,
+    Nbls,
+    Ntimes,
+    Nfreqs,
+    Nsrcs,
+    quiet,
+):
+    try:
+        import tqdm
+    except ImportError as err:
+        raise ImportError(
+            "The tqdm module must be installed to use a tqdm progress bar."
+        ) from err
+
+    mpi.start_mpi()
+    rank = mpi.get_rank()
+    comm = mpi.get_comm()
+    Npus = mpi.get_Npus()
+    status = mpi.get_status()
+    if Npus < 2:
+        raise ValueError(
+            "At least 2 PUs are required to run with tqdm backend."
+        )
+
+    uvdata_indices = []
+
+    if rank == 0:
+        completed_workers = 0
+        n_workers = Npus - 1
+        with tqdm.tqdm(total=Ntasks_tot, unit="UVTask", leave=True) as pbar:
+            while completed_workers < n_workers:
+                msg = comm.recv(source=mpi.MPI.ANY_SOURCE, tag=mpi.MPI.ANY_TAG, status=status)
+                tag = status.Get_tag()
+
+                if tag == mpi.tags.READY:
+                    pass
+                elif tag == mpi.tags.DONE:
+                    uv_container.data_array[msg[1]] += msg[0]
+                    pbar.update(1)
+
+                elif tag == mpi.tags.EXIT:
+                    completed_workers += 1
+                else:
+                    raise ValueError(f"{msg} {tag}")
```

Comment:
There's two lines in here that just _won't_ ever get touched by tests. the `READY` tag is never sent back from a worker thread, and I think the else catch was a debug thing that just never went away just in case. Ideally the "enum" defined in the mpi module is an exhaustive list of statuses that can be sent back. Would it be preferred to just take out these lines?